### DISCUSSION TITLE
Multiple Input Form Validation

### DISCUSSION THREAD

**Discussion by natankeddem** [zauberzeug/nicegui#1791] [General]
### Question

There have been several discussions in the past about validating multiple `ui.input` elements with a `ui.button` as a submission mechanism. In this interaction you don't want to enable the `ui.button` until all required `ui.input` elements are properly validated. I was able to make this work with a fairly complex class override of both elements. I didn't see anything obvious to make this work more easily but maybe there is a simpler more elegant solution. Anyway onto some most likely over complicated code...
```py
from typing import Any, Callable, Dict, List, Optional
from nicegui import ui
from nicegui.elements.mixins.value_element import ValueElement


class MyButton(ui.button):
    def __init__(
        self, text: str = "", *, on_click: Callable[..., Any] | None = None, color: str | None = "primary", icon: str | None = None
    ) -> None:
        super().__init__(text, on_click=on_click, color=color, icon=icon)
        self.enablers = {}

    def enabler(self, func):
        def wrapper(*args, **kwargs):
            value = func(*args, **kwargs)
            self.enablers[args[1].id] = value
            error_found = False
            for enabler in self.enablers.values():
                if enabler is not True:
                    error_found = True
                    break
            if error_found is True:
                self.disable()
            else:
                self.enable()
            return value

        return wrapper


class MyInput(ui.input):
    def __init__(
        self,
        label: str | None = None,
        *,
        placeholder: str | None = None,
        value: str = "",
        password: bool = False,
        password_toggle_button: bool = False,
        on_change: Callable[..., Any] | None = None,
        autocomplete: List[str] | None = None,
        validation: Dict[str, Callable[..., bool]] = {},
    ) -> None:
        super().__init__(
            label,
            placeholder=placeholder,
            value=" ",
            password=password,
            password_toggle_button=password_toggle_button,
            on_change=on_change,
            autocomplete=autocomplete,
            validation=validation,
        )
        self.value = ""

    def on_value_change(self, value: Any) -> None:
        ValueElement.on_value_change(self, value)
        for message, check in self.validation.items():
            if not check(value, self):
                self._error = message
                self.props(f'error error-message="{message}"')
                break
        else:
            self._error = None
            self.props(remove="error")


col = ui.column()
row = ui.row()
with row:
    button = MyButton("C")


@button.enabler
def is_too_short(v, e):
    return v and len(v) > 0


with col:
    a = MyInput("A", validation={"Too short!": lambda v, e: is_too_short(v, e)})
    b = MyInput("B", validation={"Too short!": lambda v, e: is_too_short(v, e)})

ui.run()
```

Alternate custom logic enabler class solution:
```py
from nicegui import ui

class Enable:
    def __init__(self) -> None:
        self.inputs = {}
        self.no_errors = True

    def is_too_short(self, v):
        return v and len(v) > 0

    def on_change(self, e):
        valid = self.is_too_short(e.value)
        self.inputs[e.sender.id] = valid
        self.update()

    def update(self):
        for i in self.inputs.values():
            if i is not True:
                self.no_errors = False
                break
        else:
            self.no_errors = True


enable = Enable()
a = ui.input("A", value=" ", on_change=enable.on_change, validation={"Too short!": enable.is_too_short})
b = ui.input("B", value=" ", on_change=enable.on_change, validation={"Too short!": enable.is_too_short})
c = ui.button("C")
a.value = ""
b.value = ""
c.bind_enabled_from(enable, "no_errors")
ui.run()
```

I am not sure which I like better at the moment.

**Comment by rodja** [comment]
Thanks for the code examples @natankeddem. I think https://github.com/zauberzeug/nicegui/discussions/801 and https://github.com/zauberzeug/nicegui/discussions/1242 are related. I find your second version more compact and simpler to understand. Why do you need to derive MyInput in the first version?

**Reply by natankeddem** [reply]
I thought you would like the 2nd version; it is a modification of your example https://github.com/zauberzeug/nicegui/discussions/1268#discussioncomment-6558120. The reason for deriving `MyInput` was to not use the `on_change` so I needed the validation to provide me with access to the element itself to get an id, not just the value. I use the id to automatically index all elements that execute any validation function that is decorated.

**Comment by falkoschindler** [comment]
I also like the second solution, @natankeddem. 🙂

I thought I might find some shortcuts and ended up with the following `ErrorChecker`:

```py
class ErrorChecker:
    def __init__(self, *elements: ValidationElement) -> None:
        self.elements = elements

    @property
    def no_errors(self) -> bool:
        return all(validation(element.value) for element in self.elements for validation in element.validation.values())

a = ui.input('A', validation={'Too short!': bool})
b = ui.input('B', validation={'Too short!': bool})
checker = ErrorChecker(a, b)
c = ui.button('C').bind_enabled_from(checker, 'no_errors')
```

Note that I simply used `bool` as validation function, since it returns `True` for all non-empty strings.
The advantage is that we don't need to deal with events.
The downside might be the binding loop in the background which constantly needs to check all validations.

**Reply by natankeddem** [reply]
I like that one @falkoschindler, you have implemented a few great tricks here that I might need to steal 😉 . I might play around with this a bit more, using your ideas. I like the idea of trying to get something that is self managing where I won't need to manually enter each input, but this will obviously have some down sides for that benefit.

On another note, while playing around with this I noticed that the validation only triggers on a change of value. The initial instaciated value doesn't seem to trigger a validation error. Is this intended?

**Reply by falkoschindler** [reply]
@natankeddem Yes, that is intended. Otherwise you would see errors in the UI before typing anything. But I have to admit, it is a bit inconvenient when implementing something like the error checker. Theoretically we could validate on instantiation but only update the UI on value change. But I'm not sure if this makes things even more complicated to follow.
